### PR TITLE
ADR: authentication and authorization

### PR DESCRIPTION
**Pull request by williamdclt**

### CODE REVIEW

File: docs/adr/01-admin-authentication.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+### Characteristics
+
+We want this authentication system to be:
+
+- Easy to set up for nestjs-admin developpers (or better: working out-of-the-box)
+- Extensible to handle a wide range of use cases
+
+### What does django-admin do?
+
+Django-admin uses the builtin "user" model, that has a username and password. It also allows to
```
Comment (TheBigBadBunce) [inline docs/adr/01-admin-authentication.md]: Allows to what? 😉

File: docs/adr/01-admin-authentication.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+3. Provide an AdminUser entity out-of-the-box, that has a username and password.
+
+   - Need to implement a login page
+   - Not very flexible. Admin entities are distinct from User entities, which is annoying (how to create admins? Choose their password?).
+   - Works entirely out-of-the-box
+
+## Decision
+
+The N.1 concern at the time is to make this library self-contained, usable out-of-the-box. For this reason, we'll go for option 1.
```
Comment (TheBigBadBunce) [inline docs/adr/01-admin-authentication.md]: Is it not option 3 as mentioned below?

File: docs/adr/01-admin-authentication.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+The downsides of this option (flexibility, Users distinct from Admins) can be mitigated:
```
Comment (TheBigBadBunce) [inline docs/adr/01-admin-authentication.md]: Why not do no. 2 instead and allow users to extend the User entity for their own purpose a la django?

File: docs/adr/01-admin-authentication.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+### What does django-admin do?
+
+Django-admin uses the builtin "user" model, that has a username and password. It also allows to
+
+### Challenges
+
+1. NestJS, as opposed to frameworks like Django, has no built-in concept of "user" with a username and password. We cannot rely on this.
+2. Users come in all shapes and form: identified through username, email, id, using a password or not... There's no invariant we can rely on.
+3. We need to make sure whatever solution we choose doesn't lock everybody out of the admin interface by default (we need to manually be able to create an admin, like django-admin with `./manage.py createsuperuser`)
```
Comment (joshwarwick) [inline docs/adr/01-admin-authentication.md]: This is a super good point

File: docs/adr/01-admin-authentication.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+### Options
+
+1. Have the developper extend the Admin controller, which will allow them to define their own guard and authentication logic
```
Comment (joshwarwick) [inline docs/adr/01-admin-authentication.md]: typo on "developper"

File: docs/adr/01-admin-authentication.md (hunk 2 above)
Comment (joshwarwick) [reply docs/adr/01-admin-authentication.md]: N.3 ;)

File: docs/adr/01-admin-authentication.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+The N.1 concern at the time is to make this library self-contained, usable out-of-the-box. For this reason, we'll go for option 1.
+
+The downsides of this option (flexibility, Users distinct from Admins) can be mitigated:
+
+- For more flexibility, we could also provide a AdminUserInterface, that can be implemented by another entity (a User entity very probably) and allow configuring the AdminModule to use this entity (option 2)
+- We can still let the developper extend the admin controller to implement their own guard (option 1., need to document how)
+
+## Consequences / roadmap
+
+The very first step is to implement option 3 (login page, AdminUser entity, authentication logic). This will make this library valuable. This is a high priority, Theodo developpers are assigned to it.
```
Comment (joshwarwick) [inline docs/adr/01-admin-authentication.md]: type two ps again lol

File: docs/adr/01-admin-authentication.md (hunk 1 above)
Comment (williamdclt) [inline docs/adr/01-admin-authentication.md]:
```suggestion
Django-admin uses the builtin "user" model, that has a username and password. It allows them to rely on these fields for authentication, in addition to `is_staff` and `is_superuser` fields.
```

File: docs/adr/01-admin-authentication.md (hunk 5 above)
Comment (williamdclt) [inline docs/adr/01-admin-authentication.md]:
```suggestion
1. Have the developer extend the Admin controller, which will allow them to define their own guard and authentication logic
```

File: docs/adr/01-admin-authentication.md (hunk 3 above)
Comment (williamdclt) [reply docs/adr/01-admin-authentication.md]: Tempting, but it means the lib wouldn't work out-of-the-box, you need to touch your user model before (which might not exist if you're just starting a project)

**Pull request merged**